File: ui/src/types/analysis.ts
export interface KpiMetrics {
  total_orders: number
  avg_delivery_time: number
  peak_delivery_time: number
  non_peak_delivery_time: number
}

export interface PeakHourRow {
  hour: number
  avg_delivery_duration: number
  is_peak: boolean
}

export interface CityDemandRow {
  city: string
  total_orders: number
}

export interface MlResult {
  model: string
  features: string[]
  target: string
  mae: number
  r2: number
  test_size: number
  random_seed: number
}

export interface SqlResults {
  delivery_by_hour: Record<string, unknown>[]
  peak_vs_non_peak: Record<string, unknown>[]
  city_performance: Record<string, unknown>[]
}

export interface ResultMeta {
  row_count: number
  columns: string[]
  duration_ms: number
  generated_at: string
}

export interface AnalysisResult {
  meta: ResultMeta
  kpis: KpiMetrics
  peak_hour_trend: PeakHourRow[]
  city_demand: CityDemandRow[]
  ml: MlResult
  sql: SqlResults
}

export interface ApiError {
  code: string
  message: string
}

File: api/schemas/analysis.py
from pydantic import BaseModel


class KpiMetrics(BaseModel):
    total_orders: int
    avg_delivery_time: float
    peak_delivery_time: float
    non_peak_delivery_time: float


class PeakHourRow(BaseModel):
    hour: int
    avg_delivery_duration: float
    is_peak: bool


class CityDemandRow(BaseModel):
    city: str
    total_orders: int


class MlResult(BaseModel):
    model: str
    features: list[str]
    target: str
    mae: float
    r2: float
    test_size: float
    random_seed: int


class SqlRow(BaseModel):
    model_config = {"extra": "allow"}


class SqlResults(BaseModel):
    delivery_by_hour: list[dict]
    peak_vs_non_peak: list[dict]
    city_performance: list[dict]


class ResultMeta(BaseModel):
    row_count: int
    columns: list[str]
    duration_ms: int
    generated_at: str


class AnalysisResult(BaseModel):
    meta: ResultMeta
    kpis: KpiMetrics
    peak_hour_trend: list[PeakHourRow]
    city_demand: list[CityDemandRow]
    ml: MlResult
    sql: SqlResults
